src/aegis/src/aegis_server.erl

```diff
@@ -0,0 +1,338 @@
+            lists:foreach(fun(#{from := From}) ->
+                gen_server:reply(From, {error, Error})
+            end, WaitList),
+            NewSt = St#{waiters := Waiters1, unwrappers := Unwrappers2},
+            {noreply, NewSt, ?TIMEOUT}
+    end;
+
+handle_info(_Msg, St) ->
+    {noreply, St}.
+
+
+code_change(_OldVsn, St, _Extra) ->
+    {ok, St}.
+
+
+%% workers functions
+
+do_generate_key(#{} = Db, Options) ->
+    process_flag(sensitive, true),
+    try
+        aegis_key_manager:generate_key(Db, Options)
+    of
+        Resp ->
+            exit(Resp)
+    catch
+        _:Error ->
+            exit({error, Error})
+    end.
+
+
+do_unwrap_key(#{aegis := AegisConfig} = Db) ->
+    process_flag(sensitive, true),
+    try
+        aegis_key_manager:unwrap_key(Db, AegisConfig)
+    of
+        Resp ->
+            exit(Resp)
+    catch
+        _:Error ->
+            exit({error, Error})
+    end.
+
+
+do_encrypt(From, DbKey, #{uuid := UUID}, Key, Value) ->
+    process_flag(sensitive, true),
+    try
+        EncryptionKey = crypto:strong_rand_bytes(32),
+        <<WrappedKey:320>> = aegis_keywrap:key_wrap(DbKey, EncryptionKey),
+
+        {CipherText, <<CipherTag:128>>} =
+            ?aes_gcm_encrypt(
+               EncryptionKey,
+               <<0:96>>,
+               <<UUID/binary, 0:8, Key/binary>>,
+               Value),
+        <<1:8, WrappedKey:320, CipherTag:128, CipherText/binary>>
+    of
+        Resp ->
+            gen_server:reply(From, Resp)
+    catch
+        _:Error ->
+            gen_server:reply(From, {error, Error})
+    end.
+
+
+do_decrypt(From, DbKey, #{uuid := UUID}, Key, Value) ->
+    process_flag(sensitive, true),
+    try
+        case Value of
+            <<1:8, WrappedKey:320, CipherTag:128, CipherText/binary>> ->
+                case aegis_keywrap:key_unwrap(DbKey, <<WrappedKey:320>>) of
+                    fail ->
+                        erlang:error(decryption_failed);
+                    DecryptionKey ->
+                        Decrypted =
+                        ?aes_gcm_decrypt(
+                            DecryptionKey,
+                            <<0:96>>,
+                            <<UUID/binary, 0:8, Key/binary>>,
+                            CipherText,
+                            <<CipherTag:128>>),
+                        if Decrypted /= error -> Decrypted; true ->
+                            erlang:error(decryption_failed)
+                        end
+                end;
+            _ ->
+                erlang:error(not_ciphertext)
+        end
+    of
+        Resp ->
+            gen_server:reply(From, Resp)
+    catch
+        _:Error ->
+            gen_server:reply(From, {error, Error})
+    end.
+
+
+%% private functions
+
+maybe_spawn_worker(St, From, Action, Db, Key, Value) ->
+    #{
+        cache := Cache,
+        waiters := Waiters
+    } = St,
+
+    #{aegis := AegisConfig} = Db,
+
+    case lookup(Cache, AegisConfig) of
```

Comment:
using the entire config as the key in ets table highlights this opaque-but-not-really problem.

Reply:
Agree. Let's use database's UUID as a key.